### ISSUE TITLE
Deprecation warnings are not shown for code in build.sbt

### ISSUE THREAD

**Issue by dwijnand** [sbt/sbt#2783]
## steps

**build.sbt**

``` scala
@deprecated("Don't use", "1.0.0")
def getName() = "foo"

name := getName()
```

**project/plugins.sbt**

``` scala
scalacOptions += "-deprecation"
```

```
17:49:58 $ sbt
[info] Loading global plugins from /Users/dnw/.dotfiles/.sbt/0.13/plugins
[info] Loading project definition from /s/t/project
[info] Set current project to foo (in build file:/s/t/)
```
## problem

No deprecation warnings are shown.
## expectation

They should be shown.
## notes

sbt.version=0.13.12

Fixing this bug is particularly important for the impending 0.13.13 release where the old fishbone operators are deprecated (#2716) and the tuple enrichments are deprecated (#2764).

**Comment by eed3si9n** [comment]
The compilation of the build.sbt DSL fragments are done in https://github.com/sbt/sbt/blob/0.13/main/actions/src/main/scala/sbt/compiler/Eval.scala, I think.
First try hard coding "-deprecation" at https://github.com/sbt/sbt/blob/0.13/main/actions/src/main/scala/sbt/compiler/Eval.scala#L52. That didn't work for me.
Once that works, we should be able to pass "-deprecation" from Load.scala (https://github.com/sbt/sbt/blob/0.13/main/src/main/scala/sbt/Load.scala#L270-L274) and Defaults.scala.

``` scala
  // build.sbt is treated a Scala source of metabuild, so to enable deprecation flag on build.sbt we set the option here.
  lazy val deprecationSettings: Seq[Setting[_]] =
    inConfig(Compile)(Seq(
      scalacOptions := {
        val old = scalacOptions.value
        val existing = old.toSet
        val d = "-deprecation"
        if (sbtPlugin.value && !existing(d)) d :: old.toList
        else old
      }
    ))
```

**Comment by dwijnand** [comment]
Ok, so I tracked down that part of the reason that my specific example doesn't work is that sbt's parsing has to workaround https://issues.scala-lang.org/browse/SI-8859 and it does so by re-parsing the already parsed Scala code (to check if there is (are?) missing trailing bracket(s)), and it does so using position start/end, which seem to chop off the leading `@deprecated` annotation.

Moving the deprecated method to a `project/*.scala` file shows the deprecation.

**Comment by dwijnand** [comment]
Reference to position chopping and re-parsing: https://github.com/sbt/sbt/blob/v0.13.12/main/src/main/scala/sbt/internals/parser/SbtParser.scala#L148-L149

**Comment by dwijnand** [comment]
@eed3si9n Could you help me understand what you meant by from Load.scala/Defaults.scala and that code block?

**Comment by eed3si9n** [comment]
So ideally we shouldn't hard-code scalacOption flags in Eval.scala, but make it configurable as part of normal build. As I wrote in the comment within the code snippet `build.sbt` is treated as Scala source of metabuild. This means the user should be able to configure this in `project/*.sbt` if they wanted to.

**Comment by dwijnand** [comment]
I agree to making it not hard-coded, but we should default it to include (at least, for now) `-deprecation` for the meta-build (users can then opt-out with `-=`). How do I do that?

**Comment by eed3si9n** [comment]
That's what I coded above - deprecationSettings

**Comment by dwijnand** [comment]
All right, I'm just not sure where it should be added.

**Comment by dwijnand** [comment]
Fixed in #2784
